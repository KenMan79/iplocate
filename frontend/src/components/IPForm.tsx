import React, { useState } from "react";
import styled from "styled-components";
import { Input } from "./Input";
import { FaSearch } from "react-icons/fa";

interface Props {
  onSubmit(value: string): void;
}

// basic validation here, better validation on backend with ajv
// const ipRE = /^(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$/;
const validateIp = (ip: string) => {
  return !!ip;
};

export const IPForm: React.FC<Props> = (props) => {
  const { onSubmit } = props;
  const [value, setValue] = useState("");
  const [error, setError] = useState(false);

  const handleSubmit: React.FormEventHandler = (e) => {
    e.preventDefault();
    if (!validateIp(value)) {
      setError(true);
    } else {
      onSubmit(value);
      setValue("");
    }
  };

  const handleChange: React.ChangeEventHandler<HTMLInputElement> = (e) => {
    setValue(e.currentTarget.value);
    setError(false);
  };

  return (
    <>
      <Form onSubmit={handleSubmit} error={error}>
        <Input
          value={value}
          onChange={handleChange}
          placeholder="IP Address"
          aria-label="IP Address input"
        />
        <button type="submit" aria-label="Search IP Address">
          <FaSearch />
        </button>
      </Form>
      {error && <ErrorMessage>Invalid IP Address.</ErrorMessage>}
    </>
  );
};

const Form = styled.form<{ error: boolean }>`
  margin: 0;
  padding: 0;
  display: flex;
  flex-flow: row nowrap;
  align-items: stretch;
  justify-content: stretch;

  ${Input} {
    flex: 1 1 auto;
    border-right: none;
    border-top-right-radius: 0;
    border-bottom-right-radius: 0;
    border-color: ${(props) =>
      props.error ? "var(--text-error)" : "var(--text-lighter)"};
  }

  ${Input}:focus {
    outline: none;
    border-color: ${(props) =>
      props.error ? "var(--text-error)" : "var(--text-secondary)"};
    border-right: none;
  }

  button {
    flex: 0 0 auto;
    border: none;
    outline: none;
    border-top-right-radius: 5px;
    border-bottom-right-radius: 5px;
    background: var(--text-secondary);
    display: flex;
    align-items: center;
    padding: 0 1.2rem;
    font-size: 18px;
    color: #fff;
    cursor: pointer;

    svg {
      display: block;
    }
  }
`;

const ErrorMessage = styled.p`
  color: var(--text-error);
  margin: 0;
  padding: 0.5rem 0 0 0;
`;
