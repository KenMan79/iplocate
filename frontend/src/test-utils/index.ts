export * from "./test-utils";
export * from "./ipdata";
