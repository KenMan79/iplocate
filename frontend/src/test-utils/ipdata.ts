import { IP } from "types";

export const baseIP: IP = {
  continent: {
    code: "EU",
    geonameId: 6255148,
    names: {
      en: "Europe",
    },
  },
  country: {
    geonameId: 3017382,
    isInEuropeanUnion: true,
    isoCode: "FR",
    names: {
      en: "France",
    },
  },
  traits: {
    isAnonymous: false,
    isAnonymousProxy: false,
    isAnonymousVpn: false,
    isHostingProvider: false,
    isLegitimateProxy: false,
    isPublicProxy: false,
    isResidentialProxy: false,
    isSatelliteProvider: false,
    isTorExitNode: false,
    ipAddress: "2.2.2.2",
    network: "2.2.0.0/18",
  },
  city: {},
  location: {
    accuracyRadius: 500,
    latitude: 48.8582,
    longitude: 2.3387,
    timeZone: "Europe/Paris",
  },
};
