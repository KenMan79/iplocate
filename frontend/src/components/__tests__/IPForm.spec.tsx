import React from "react";
import { fireEvent, render, wait } from "test-utils/test-utils";
import { IPForm } from "../IPForm";

const submitMock = jest.fn();

afterEach(() => {
  submitMock.mockReset();
});

it("matches snapshot", () => {
  const { asFragment } = render(<IPForm onSubmit={submitMock} />);
  expect(asFragment()).toMatchSnapshot();
});

it("can enter a valid ip address", async () => {
  const { getByPlaceholderText, getByLabelText } = render(
    <IPForm onSubmit={submitMock} />
  );
  const inp = getByPlaceholderText("IP Address");
  fireEvent.change(inp, {
    target: { value: "1.1.1.1" },
  });
  await wait(() => {
    expect(inp).toHaveValue("1.1.1.1");
  });
  const submit = getByLabelText("Search IP Address");
  fireEvent.click(submit);
  await wait(() => {
    expect(submitMock).toHaveBeenCalledTimes(1);
    expect(submitMock).toHaveBeenCalledWith("1.1.1.1");
  });
});

it("does not submit invalid ip addresses", async () => {
  const invalid = ["not an ip", "555.1.1.1", "...", "1.1.1", ""];
  const {
    getByPlaceholderText,
    getByLabelText,
    getByText,
    queryByText,
  } = render(<IPForm onSubmit={submitMock} />);
  const inp = getByPlaceholderText("IP Address");
  const submit = getByLabelText("Search IP Address");

  for (let value of invalid) {
    fireEvent.change(inp, {
      target: { value },
    });
    await wait(() => {
      expect(inp).toHaveValue(value);
      expect(queryByText("Invalid IP Address.")).not.toBeInTheDocument();
    });
    fireEvent.click(submit);
    await wait(() => {
      expect(getByText("Invalid IP Address.")).toBeInTheDocument();
    });
  }
});

it("submits valid ip addresses", async () => {
  const valid = ["255.255.255.255", "0.0.0.0", "8.8.8.8", "127.0.0.1"];
  const { getByPlaceholderText, getByLabelText, queryByText } = render(
    <IPForm onSubmit={submitMock} />
  );
  const inp = getByPlaceholderText("IP Address");
  const submit = getByLabelText("Search IP Address");
  let count = 1;

  for (let value of valid) {
    fireEvent.change(inp, {
      target: { value },
    });
    await wait(() => {
      expect(inp).toHaveValue(value);
    });
    fireEvent.click(submit);
    await wait(() => {
      expect(queryByText("Invalid IP Address.")).not.toBeInTheDocument();
      expect(submitMock).toHaveBeenCalledTimes(count);
      expect(submitMock).toHaveBeenCalledWith(value);
    });
    count++;
  }
});
