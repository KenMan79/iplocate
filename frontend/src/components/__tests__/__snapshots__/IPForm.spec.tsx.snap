// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  .c2 {
  background: #fff;
  border: 2px solid var(--text-lightest);
  border-radius: 5px;
  padding: 1rem;
  font-size: 1.2rem;
  width: 100%;
  color: var(--text-primary);
  font-family: var(--font-family-mono);
}

.c0 {
  margin: 0;
  padding: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: row nowrap;
  -ms-flex-flow: row nowrap;
  flex-flow: row nowrap;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-box-pack: stretch;
  -webkit-justify-content: stretch;
  -ms-flex-pack: stretch;
  justify-content: stretch;
}

.c0 .c1 {
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  border-right: none;
  border-top-right-radius: 0;
  border-bottom-right-radius: 0;
  border-color: var(--text-lighter);
}

.c0 .c1:focus {
  outline: none;
  border-color: var(--text-secondary);
  border-right: none;
}

.c0 button {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  border: none;
  outline: none;
  border-top-right-radius: 5px;
  border-bottom-right-radius: 5px;
  background: var(--text-secondary);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0 1.2rem;
  font-size: 18px;
  color: #fff;
  cursor: pointer;
}

.c0 button svg {
  display: block;
}

<form
    class="c0"
  >
    <input
      aria-label="IP Address input"
      class="c1 c2"
      placeholder="IP Address"
      value=""
    />
    <button
      aria-label="Search IP Address"
      type="submit"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
        />
      </svg>
    </button>
  </form>
</DocumentFragment>
`;
